### PR TITLE
Remove /ui/ redirect

### PR DESCRIPTION
**Pull request by liggitt**
The existing kube-apiserver hard-codes `/ui` to redirect to an optional add-on, which is not appropriate. It does not work in the following scenarios:
* https-enabled dashboards
* the dashboard is deployed to a different namespace or service name
* the dashboard is not installed at all
* authorization is enabled and does not allow access to /ui

This PR removes the hard-coded `/ui` redirect.

```release-note
apiserver: the /ui kube-dashboard redirect has been removed. Follow instructions specific to your deployment to access kube-dashboard
```

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-cli-pr-reviews

File: pkg/master/master.go [hunk 1]
```diff
@@ -270,7 +269,7 @@ func (cfg *Config) Complete(informers informers.SharedInformerFactory) Completed
 	}
 
 	// enable swagger UI only if general UI support is on
```
Comment (dixudx) [inline pkg/master/master.go]: Remove this comment?

**Comment by maciaszczykm** [comment]
I have opened #53046 to fix this issue without `/ui` removal. It is not merged yet because it is blocked by #53382. This will decrease Dashboard discoverability a lot. 

@kubernetes/dashboard-maintainers

**Comment by sttts** [comment]
I would also prefer to keep the redirect if we manage to fix it. As this is a public interface, we would have to deprecate that anyway, e.g. in the changelogs.

**Comment by liggitt** [comment]
> I would also prefer to keep the redirect if we manage to fix it

I don't think the redirect is appropriate... the API server shouldn't assume the presence of a particular add-on like kube-dashboard.

**Comment by floreks** [comment]
> I would also prefer to keep the redirect if we manage to fix it

We can easily fix the logic and make redirect smarter (similar to logic in this PR). Many people got used to it and are often using it to access Dashboard. We can see that based on number of issues that mention this redirect.

**Comment by maciaszczykm** [comment]
+1 @floreks

**Comment by liggitt** [comment]
> We can easily fix the logic and make redirect smarter (similar to logic in this PR). Many people got used to it and are often using it to access Dashboard. We can see that based on number of issues that mention this redirect.

kube-dashboard is not part of core, it's an add-on. I don't think it is appropriate to hardcode UI-oriented redirects for optional add-ons into the API server

**Comment by ericchiang** [comment]
We've had customers notice issues like this in the past, e.g. `kubectl top` breaking because we didn't ship heapster by default. If a customer sees something that looks broken, that reflects badly on us, even if it was because we chose not to ship an optional add-on.

+1 to removing this. The fewer add-ons core assumes are running, the better.

**Comment by liggitt** [comment]
moved the cluster-info change to https://github.com/kubernetes/kubernetes/pull/53897 and limited this to the UI redirect change

**Comment by liggitt** [comment]
assigned to kube-apiserver owners

**Comment by lavalamp** [comment]
I can understand deprecating this but I don't think we can remove it without warning.

**Comment by lavalamp** [comment]
I would probably start by sending an email to kubernetes-dev and see what happens.

**Comment by bryk** [comment]
Updating documentation to remove all `/ui` references is a prerequisite to this, I think. We can replace the `/ui` references with direct service proxy link.

**Comment by liggitt** [comment]
> Updating documentation to remove all /ui references is a prerequisite to this, I think. We can replace the /ui references with direct service proxy link.

There are no references to `/ui` in the kubernetes website. The only documentation reference I found in the kubernetes org was in kubernetes/dashboard. Opened https://github.com/kubernetes/dashboard/pull/2620 to update that.

**Comment by cdenneen** [comment]
Why can't the redirect just redirect to new URL?

**Comment by liggitt** [comment]
@bryk @floreks updated to remove /ui/ redirect as discussed in https://github.com/kubernetes/kubernetes/pull/53046#discussion_r153261279

**Pull request merged**